### PR TITLE
feat(`forge`): add createWallet() cheat code for new Wallet struct

### PR DESCRIPTION
**Pull request by trevorgjohnson**
## Motivation
Issue https://github.com/foundry-rs/foundry/issues/4790

## Solution

Add the struct `Wallet` to contain its `privateKey`, `addr`, `publicKeyX`, and `publicKeyY`.
`createWallet` was added to create `Wallet`'s and `sign`, and `getNonce` were also added to as helpers for `Wallet`

### CODE REVIEW

File: evm/src/executor/inspector/cheatcodes/env.rs [hunk 1]
```diff
@@ -503,7 +503,7 @@ pub fn apply<DB: DatabaseExt>(
                 Ok(Bytes::new())
             },
         )??,
-        HEVMCalls::GetNonce(inner) => {
+        HEVMCalls::GetNonce1(inner) => {
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/env.rs:506]: Because of the placement of `getNonce(Wallet)` in _abi/abi/HEVM.sol_, `getNonce(address)` was renamed to `HEVMCalls::GetNonce1`

File: evm/src/executor/inspector/cheatcodes/env.rs [hunk 2]
```diff
@@ -519,6 +519,22 @@ pub fn apply<DB: DatabaseExt>(
...
+                state,
+            )?;
+
+            // TODO:  this is probably not a good long-term solution since it might mess up the gas
+            // calculations
+            data.journaled_state.load_account(h160_to_b160(inner.0.addr), data.db)?;
+
+            // we can safely unwrap because `load_account` insert inner.0 to DB.
+            let account = data.journaled_state.state().get(&h160_to_b160(inner.0.addr)).unwrap();
+            abi::encode(&[Token::Uint(account.info.nonce.into())]).into()
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/env.rs:536]: Should all be the same as `HEVMCalls::GetNonce1` except `.get(&h160_to_b160(inner.0.addr))`

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 3]
```diff
@@ -120,6 +124,24 @@ fn sign(private_key: U256, digest: H256, chain_id: U256) -> Result {
...
+fn create_wallet(private_key: U256, label: Option<String>, state: &mut Cheatcodes) -> Result {
+    let key = parse_private_key(private_key)?;
+    let addr = utils::secret_key_to_address(&key);
+
+    let pub_key = key.verifying_key().as_affine().to_encoded_point(false);
+    let pub_key_x = pub_key.x().ok_or("No x coordinate was found")?;
+    let pub_key_y = pub_key.y().ok_or("No y coordinate was found")?;
+
+    let pub_key_x = U256::from(pub_key_x.as_slice());
+    let pub_key_y = U256::from(pub_key_y.as_slice());
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/util.rs:136]: This is the best way I could find to extract the x and y coordinates. I'm not extremely familiar with the details of ECDSA algorithms, so I'm not 100% sure if this is accurate either.

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 4]
```diff
@@ -120,6 +124,24 @@ fn sign(private_key: U256, digest: H256, chain_id: U256) -> Result {
...
+
+    if let Some(label) = label {
+        state.labels.insert(addr, label);
+    }
+
+    Ok((addr, pub_key_x, pub_key_y, private_key).encode().into())
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/util.rs:142]: Ideally, `private_key` or `addr` should be reconstructed with `pub_key_x` and `pub_key_y` to ensure that the coordinates are correct, but I couldn't find a great way do that yet. I may make another commit adding that and reference that commit here, but I'm currently looking for suggestions 👍

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 5]
```diff
@@ -222,7 +244,15 @@ pub fn apply<DB: Database>(
...
-        HEVMCalls::Sign(inner) => sign(inner.0, inner.1.into(), data.env.cfg.chain_id.into()),
+        HEVMCalls::Sign0(inner) => sign(inner.0, inner.1.into(), data.env.cfg.chain_id.into()),
+        HEVMCalls::CreateWallet0(inner) => {
+            create_wallet(U256::from(keccak256(&inner.0)), Some(inner.0.clone()), state)
+        }
+        HEVMCalls::CreateWallet1(inner) => create_wallet(inner.0, None, state),
+        HEVMCalls::CreateWallet2(inner) => create_wallet(inner.0, Some(inner.1.clone()), state),
+        HEVMCalls::Sign1(inner) => {
+            sign(inner.0.private_key, inner.1.into(), data.env.cfg.chain_id.into())
+        }
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/util.rs:255]: Just wanted to compile a quick list to explain the functionality of each changed line here:

- `Sign0`: the same as the old `Sign`. Just needed to be renamed for new `function sign(Wallet)`
- `CreateWallet0`: `function createWallet(string)` to derive private key and label wallet with `string`
- `CreateWallet1`: `function createWallet(uint256)` create new wallet with private key `uint256` _(no label)_
- `CreateWallet2`: `function createWallet(uint256,string)` create new wallet with private key `uint256` and label with `string`
- `Sign1`: `function sign(Wallet,bytes32)` but uses wallet's private key

File: testdata/cheats/Wallet.t.sol [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    function testCreateWalletStringPrivAndLabel() public {
+        bytes memory privKey = "this is a priv key";
+        Cheats.Wallet memory wallet = cheats.createWallet(string(privKey));
+
+        address expectedAddr = cheats.addr(wallet.privateKey);
+        assertEq(expectedAddr, wallet.addr);
+
+        string memory label = cheats.getLabel(wallet.addr);
+        assertEq(label, string(privKey), "labelled address != wallet.addr");
+    }
```
Comment (trevorgjohnson) [inline testdata/cheats/Wallet.t.sol:21]: Ideally, there'd be another section of this test that checks the validity of `wallet.publicKeyX` and `wallet.publicKeyY`, but I wasn't sure what the best way of doing that was

**Comment by mds1** [comment]
@pmerkleplant what was your original use case for pubkeys? Mainly asking because there's three ways we can represent them, and I'm wondering if we should choose one and you can convert between them as needed (maybe tricky in solidity), or just give all 3 in the `Wallet` struct (but that feels cluttered)

```solidity
// Approach 1
uint256 publicKeyX;
uint256 publicKeyY;

// Approach 2
bytes publicKey; // 0x04{x}{y}

// Approach 3;
bytes compressedPublicKey; // 0x02{x} or 0x03{x}
```

**Comment by pmerkleplant** [comment]
> @pmerkleplant what was your original use case for pubkeys? 

Our use case is to aggregate public keys, i.e. perform elliptic curve point addition. We also have to derive the Ethereum address of the resulting public key. 

We hope that the cheatcode makes our fuzzing test suite faster, in which we receive sets of random private keys and compute and aggregate their public keys.

Therefore, the most suitable approach for us would be the first/currently implemented one, i.e. receiving the x and y Affine coordinates as `uint`s.

I would personally argue against approach 3 as the y coordinate would need to be computed again, making even deriving the Ethereum address computation heavy and error-prone.

@mds1 Out of curiosity, could you elaborate on where approach 2 is coming from? Is this a format standardized somewhere?

@trevorgjohnson Many thanks for the work so far!

File: testdata/cheats/Wallet.t.sol (hunk 6 above)
Comment (pmerkleplant) [reply testdata/cheats/Wallet.t.sol:21]: One way would be to verify the address derived from the coordinates equals the address derived via the `addr(unit privKey)(address)` cheatcode.

Deriving the Ethereum address is specified in the Yellow Paper "Appendix F: Signing Transactions" §134. A possible Solidity implementation is:
```solidity
function addressOf(uint xCoordinate, uint yCoordinate) internal pure returns (address) {
    return address(uint160(keccak256(abi.encode(xCoordinate, yCoordinate))));
}
```

**Comment by mds1** [comment]
> Therefore, the most suitable approach for us would be the first/currently implemented one, i.e. receiving the x and y Affine coordinates as `uint`s.
> I would personally argue against approach 3 as the y coordinate would need to be computed again, making even deriving the Ethereum address computation heavy and error-prone.

Got it, thanks. I have a use case as part of ERC-5564/ERC-6538 where we may store compressed public keys in a registry, but starting with the x/y coordinates and compressing/uncompressing in solidity shouldn't be too bad, so I'd be ok with moving forward with the current approach for the `Wallet` struct

> Out of curiosity, could you elaborate on where approach 2 is coming from? Is this a format standardized somewhere?

This is a convention from Standards for Efficient Cryptography Group where:
- uncompressed pubkeys are 65 bytes (0x04 prefix + 32 byte x-coordinate + 32 byte y-coordinate
- compressed pubkeys are 33 btyes (0x02 OR 0x03 prefix + 32 byte x-coordinate, where 0x02 means the y value is even, and 0x03 means it's odd

It's used by popular libs such as ethers.js and noble-secp256k1. You can read more about it and find references to more data in mastering ethereum [here](https://github.com/ethereumbook/ethereumbook/blob/develop/04keys-addresses.asciidoc#generating-a-public-key)

**Comment by trevorgjohnson** [comment]
I rebased the current version of `master

File: testdata/cheats/Wallet.t.sol [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+        // check wallet.addr against recovered address using x and y coordinates
+        expectedAddr = addressOf(wallet.publicKeyX, wallet.publicKeyY);
+        assertEq(expectedAddr, wallet.addr);
+
+        string memory label = vm.getLabel(wallet.addr);
+        assertEq(label, string(privKey), "labelled address != wallet.addr");
+    }
+
+    function testCreateWalletPrivKeyNoLabel(uint248 pk) public {
```
Comment (pmerkleplant) [inline testdata/cheats/Wallet.t.sol:32]: Instead of receiving the private key as type `uint248`, one could receive a private key seed of type `uint256` and create the private key via:
```solidity
uint internal constant Q = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141;

// Let pk ∊ [1, Q).
uint pk = bound(pkSeed, 1, Q-1);
```

This would ensure the whole eligible set of private keys is testable. (Note that `type(uint248).max < Q`, meaning the highest possible private keys are not testable)

**Review by pmerkleplant — APPROVED** [review]
Approving the Solidity part for my use case. Note that I'm not skilled to review the Rust changes.

**Comment by trevorgjohnson** [comment]
Hello @Evalir @mattsse, just wanted to check and see if you guys have had a chance to check out the newest changes to this PR yet. Thanks!

**Comment by Austinhs** [comment]
Please add :)

**Comment by Evalir** [comment]
ah whoops, we let this go stale but i think there's still interest here. @trevorgjohnson sorry! mind fixing the conflict to give this a review?

**Comment by trevorgjohnson** [comment]
> ah whoops, we let this go stale but i think there's still interest here. @trevorgjohnson sorry! mind fixing the conflict to give this a review?

No worries! The conflicts should be resolved in [d7bcc7d](https://github.com/foundry-rs/foundry/pull/5332/commits/d7bcc7d1269a98bb41a1e089cd73dbc8b5f18578)

File: testdata/cheats/Wallet.t.sol (hunk 7 above)
Comment (mds1) [reply testdata/cheats/Wallet.t.sol:32]: +1, this would be a stronger test

File: evm/src/executor/inspector/cheatcodes/env.rs [hunk 8]
```diff
@@ -523,6 +523,22 @@ pub fn apply<DB: DatabaseExt>(
...
+            correct_sender_nonce(
+                b160_to_h160(data.env.tx.caller),
+                &mut data.journaled_state,
+                &mut data.db,
+                state,
+            )?;
+
+            // TODO:  this is probably not a good long-term solution since it might mess up the gas
+            // calculations
+            data.journaled_state.load_account(h160_to_b160(inner.0.addr), data.db)?;
```
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/env.rs:536]: @mattsse wdyt here? seems ok, but might mess with gas yes

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 9]
```diff
@@ -120,6 +124,24 @@ fn sign(private_key: U256, digest: H256, chain_id: U256) -> Result {
     Ok((sig.v, r_bytes, s_bytes).encode().into())
 }
 
+fn create_wallet(private_key: U256, label: Option<String>, state: &mut Cheatcodes) -> Result {
```
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/util.rs:127]: let's add a bit of docs on what it does, even if it's a bit obvious :D

File: evm/src/executor/inspector/cheatcodes/util.rs (hunk 4 above)
Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/util.rs:142]: I think it's ok right now to construct the `addr` from the private key—is the usual way you see it implemented in most libraries, even if it should be possible to construct it from the pubkeys—constructing from the privkey is just one more step.

File: evm/src/executor/inspector/cheatcodes/util.rs (hunk 5 above)
Comment (Evalir) [reply evm/src/executor/inspector/cheatcodes/util.rs:255]: We can add these as comments, they're great!

File: testdata/cheats/Wallet.t.sol (hunk 7 above)
Comment (trevorgjohnson) [reply testdata/cheats/Wallet.t.sol]: Ha yea that's a good catch! There was an issue with `bound` not being defined, so I fixed that and this nix in this commit here: [8c36615](https://github.com/foundry-rs/foundry/pull/5332/commits/8c36615b811996e8e494e96120d3bc0f79d867ea)

File: evm/src/executor/inspector/cheatcodes/util.rs (hunk 9 above)
Comment (trevorgjohnson) [reply evm/src/executor/inspector/cheatcodes/util.rs:127]: Sounds good to me! Were you thinking about just adding some comments above this? Or just adding some `create_wallet` documentation in the foundry book?

File: evm/src/executor/inspector/cheatcodes/util.rs [hunk 10]
```diff
@@ -222,7 +248,23 @@ pub fn apply<DB: Database>(
...
-        HEVMCalls::Sign(inner) => sign(inner.0, inner.1.into(), data.env.cfg.chain_id.into()),
+        // [function sign(uint256,bytes32)] Used to sign bytes32 digests using the given private key
+        HEVMCalls::Sign0(inner) => sign(inner.0, inner.1.into(), data.env.cfg.chain_id.into()),
+        // [function createWallet(string)] Used to derive private key and label the wallet with the
+        // same string
+        HEVMCalls::CreateWallet0(inner) => {
+            create_wallet(U256::from(keccak256(&inner.0)), Some(inner.0.clone()), state)
+        }
+        // [function createWallet(uint256)] creates a new wallet with the given private key
+        HEVMCalls::CreateWallet1(inner) => create_wallet(inner.0, None, state),
```
Comment (trevorgjohnson) [inline evm/src/executor/inspector/cheatcodes/util.rs:259]: @Evalir, I looked around for an existing doc schema for each of these `match call` arms, but I couldn't find any so I made my own. Let me know what you think 👍

**Review by Evalir — APPROVED** [review]

**Comment by Evalir** [comment]
Amazing! just need forge-std PR + book update
